### ISSUE TITLE
Question about facenet classifier training

### ISSUE THREAD

**Issue by tigerhaolu727** [davidsandberg/facenet#182]
Hi, 

I was reading about the [classifier ](https://github.com/davidsandberg/facenet/wiki/Classifier-training-of-inception-resnet-v1)training code.

Just have some question about the code, and how it works.

in the classifier python file [facenet_train_classifier.py](https://github.com/davidsandberg/facenet/blob/master/src/facenet_train_classifier.py) 

there is a block of code as follow (line 111- 137):

```
        input_queue = data_flow_ops.FIFOQueue(capacity=100000,
                                    dtypes=[tf.string, tf.int64],
                                    shapes=[(1,), (1,)],
                                    shared_name=None, name=None)
                              enqueue_op=input_queue.enqueue_many([image_paths_placeholder,labels_placeholder],name='enqueue_op')
        
        nrof_preprocess_threads = 4
        images_and_labels = []
        for _ in range(nrof_preprocess_threads):
            filenames, label = input_queue.dequeue()
            images = []
            for filename in tf.unstack(filenames):
                file_contents = tf.read_file(filename)
                image = tf.image.decode_png(file_contents)
                if args.random_rotate:
                    image = tf.py_func(facenet.random_rotate_image, [image], tf.uint8)
                if args.random_crop:
                    image = tf.random_crop(image, [args.image_size, args.image_size, 3])
                else:
                    image = tf.image.resize_image_with_crop_or_pad(image, args.image_size, args.image_size)
                if args.random_flip:
                    image = tf.image.random_flip_left_right(image)
    
                #pylint: disable=no-member
                image.set_shape((args.image_size, args.image_size, 3))
                images.append(tf.image.per_image_standardization(image))
            images_and_labels.append([images, label])
```

As my understand, first we have created a FIFO queue which have shape of 1, 1:

```
        input_queue = data_flow_ops.FIFOQueue(capacity=100000,
                                    dtypes=[tf.string, tf.int64],
                                    shapes=[(1,), (1,)],
                                    shared_name=None, name=None)
```

Then on line 120, we called the following code: 

```
        filenames, label = input_queue.dequeue()
```

As my understanding, the dequeue operation will only return the single object so

filenames =  tensor of [image_file_path of one image]
label = class_identification

Then we called

```
            images = []
            for filename in tf.unstack(filenames):
                file_contents = tf.read_file(filename)
                image = tf.image.decode_png(file_contents)
                if args.random_rotate:
                    image = tf.py_func(facenet.random_rotate_image, [image], tf.uint8)
                if args.random_crop:
                    image = tf.random_crop(image, [args.image_size, args.image_size, 3])
                else:
                    image = tf.image.resize_image_with_crop_or_pad(image, args.image_size, args.image_size)
                if args.random_flip:
                    image = tf.image.random_flip_left_right(image)
    
                #pylint: disable=no-member
                image.set_shape((args.image_size, args.image_size, 3))
                images.append(tf.image.per_image_standardization(image))
            images_and_labels.append([images, label])
```

First question, if tf.unstack returns a list of one item, the images will looks like [[rgb_of_the_image]], then 

images_and_labels will contains data like [[[rgb_of_the_image], label_1],[[rgb_of_the_image], label_2]] ([rgb_of_the_image] = array of rgb_of_the_image)

how would this work, when we pass it into the training session, is that the ideal training data should be as follow: 

[[ rgb_of_the_image, label_1],[rgb_of_the_image, label_2]]

If I didn't explain my question clear, please let me know, Appreciate on any feedback given.

Thanks

**Comment by davidsandberg** [comment]
Yes, essentially you are right. A batch of examples is created by the call to tf.train.batch_join and that is then fed to the model.
